pkg/encrypt/server-side.go

```diff
@@ -60,139 +52,229 @@ type PBKDF func(password, salt []byte) ServerSide
+func (sses3) MarshalCOPY(h http.Header, op CopyOp) {
+	if op == CopyTarget {
+		h.Add(amzServerSideEncryption, "AES256")
 	}
-	return kms{key: keyID, context: serializedContext, hasContext: true}, nil
 }
 
-// NewSSEC returns a new server-side-encryption using SSE-C and the provided key.
-// The key must be 32 bytes long.
-func NewSSEC(key []byte) (ServerSide, error) {
-	if len(key) != 32 {
-		return nil, errors.New("encrypt: SSE-C key must be 256 bit long")
-	}
-	sse := ssec{}
-	copy(sse[:], key)
-	return sse, nil
+type ssekms struct {
+	keyID   string
+	context []byte
 }
 
-// SSE transforms a SSE-C copy encryption into a SSE-C encryption.
-// It is the inverse of SSECopy(...).
+// WithKMS encrypts the object with the key referenced by keyID
+// that is managed a KMS.
 //
-// If the provided sse is no SSE-C copy encryption SSE returns
-// sse unmodified.
-func SSE(sse ServerSide) ServerSide {
-	if sse == nil || sse.Type() != SSEC {
-		return sse
-	}
-	if sse, ok := sse.(ssecCopy); ok {
-		return ssec(sse)
+// The context, if not nil, is bound to the encrypted object
+// and must be provided whenever the object should be accessed
+// again.
+func WithKMS(keyID string, context map[string]string) (ServerSide, error) {
+	var (
+		bytes []byte
+		err   error
+	)
+	if context != nil {
+		bytes, err = json.Marshal(context)
+		if err != nil {
+			return nil, err
+		}
 	}
-	return sse
+	return ssekms{
+		keyID:   keyID,
+		context: bytes,
+	}, nil
 }
 
-// SSECopy transforms a SSE-C encryption into a SSE-C copy
-// encryption. This is required for SSE-C key rotation or a SSE-C
-// copy where the source and the destination should be encrypted.
-//
-// If the provided sse is no SSE-C encryption SSECopy returns
-// sse unmodified.
-func SSECopy(sse ServerSide) ServerSide {
-	if sse == nil || sse.Type() != SSEC {
-		return sse
+// String returns "SSE-KMS" as the string representation
+// of SSE-KMS.
+func (s ssekms) String() string { return "SSE-KMS" }
+
+// MarshalPUT adds the SSE-KMS server-side encryption HTTP
+// headers.
+func (s ssekms) MarshalPUT(h http.Header) {
+	h.Add(amzServerSideEncryption, "amz:kms")
+	if s.keyID != "" {
+		h.Add(amzServerSideEncryptionKmsKeyID, s.keyID)
 	}
-	if sse, ok := sse.(ssec); ok {
-		return ssecCopy(sse)
+	if s.context != nil {
+		h.Add(amzServerSideEncryptionContext, base64.StdEncoding.EncodeToString(s.context))
 	}
-	return sse
 }
 
-type ssec [32]byte
-
-func (s ssec) Type() Type { return SSEC }
-
-func (s ssec) Marshal(h http.Header) {
-	keyMD5 := md5.Sum(s[:])
-	h.Set(sseCustomerAlgorithm, "AES256")
-	h.Set(sseCustomerKey, base64.StdEncoding.EncodeToString(s[:]))
-	h.Set(sseCustomerKeyMD5, base64.StdEncoding.EncodeToString(keyMD5[:]))
+// MarshalGET adds the SSE-KMS server-side encryption HTTP
+// header containing the encryption context, if any.
+func (s ssekms) MarshalGET(h http.Header) {
+	if s.context != nil {
+		h.Add(amzServerSideEncryptionContext, base64.StdEncoding.EncodeToString(s.context))
+	}
```

Comment:
was looking at headers needed for kms in aws GetObject - the encryption context doesn't need to be specified on GET per [this doc](https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html). Is this needed by KES?

Reply:
See: 
https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#encrypt_context

> When you include an encryption context in an encryption request, it is cryptographically bound to the ciphertext such that the same encryption context is required to decrypt (or decrypt and re-encrypt) the data. If the encryption context provided in the decryption request is not an exact, case-sensitive match, the decrypt request fails. Only the order of the key-value pairs in the encryption context can vary.

Reply:
S3 uses the AWS-KMS. AWS-KMS requires the context to decrypt a ciphertext (data encryption key) since the context is crypto. bound to the ciphertext.
Therefore, S3 GET / HEAD / PUT / COPY operations may require the context.

Reply:
ok